### DISCUSSION TITLE
Resolve the storage dilemma under Android 11 by using "All files access"

### DISCUSSION THREAD

**Discussion by sonora** [osmandapp/OsmAnd#13538] [Ideas]
Let's use "All files access" for OsmAnd:

https://developer.android.com/training/data-storage/manage-all-files#all-files-access-google-play says: "You should request the MANAGE_EXTERNAL_STORAGE permission only when your app cannot effectively make use of the more privacy-friendly APIs, such as Storage Access Framework or the Media Store API."

I think we have collected multiple issues proving thar this appplies to us.

As I suggested in https://github.com/osmandapp/OsmAnd/issues/13254#issuecomment-1007243270,  under the circumstances we should simply use the Android "All files access" permission reserved for exceptional cases (https://developer.android.com/training/data-storage/manage-all-files#all-files-access-google-play) and see if the Play store has any objections at all?

Reasoning is that OsmAnd is essentially a "Document management app" for maps, tracks, locations, travel-wikis, trip recordings, etc.

**Comment by vshcherb** [comment]
Good that you split that discussion. Quote from previous discussion
> I wouldn't do it we are definitely not in this business and neither of our competitors has this permission. Of course we could distribute this version via other markets but for sure not via GPlay. I've been having blocked OsmAnd for 4 times and that's not a pleasant experience even if user could continue download it.

I think if that's stays critical we could consider implementing "Hot cache approach" i.e. copy maps from cold storage to hot storage by request. We know how to copy maps via SAF (it's slow but it's working) user can give permission to any folder to any app.

We're not going to implement SAF's approach to rendering cause it's slow anyway and there are no visible guarantees we can speed it up

**Comment by kevinmcmurtrie** [comment]
Wouldn't the hot cache cause stalls again?  There are plenty of maps over 1GB that would be at least 12 seconds to copy even without SAF in the way.  Double that with contour lines.  If I'm scrolling around to plan a vacation, do I deplete internal storage?

**Reply by vshcherb** [reply]
yes it would. Probably we will limit hot cache up to 1-4 GB which is enough for 10-40 maps and then progress will be displayed that files are swapped.

**Reply by kevinmcmurtrie** [reply]
US California map + contour + hillshade totals 2.5 GB and that's hardly the largest set.

**Reply by vshcherb** [reply]
Unfortunately yes with slopes/srtm it's a different story... Whole world maps though is only 160-200 GB

**Comment by kevinmcmurtrie** [comment]
You could link to https://issuetracker.google.com/issues/179412245 in your request to Google to use MANAGE_EXTERNAL_STORAGE.  I used OsmAnd+ as an example backing up that issue because it's a paid app that hasn't bailed to F-Droid yet.

Googlers believe they can do no wrong so they've been rejecting SAF tickets like crazy. "It's for security," they said while collecting private details from millions of people to use for manipulation. Good luck.

**Reply by vshcherb** [reply]
I think we should not try with this API cause it doesn't look designed for us. Also I did a small improvement which should be pretty significant for initialization on SAF. 
I think Googlers we will win in the end and new phones won't have SDCards any more and probably won't exist in 2-3 years.

**Reply by kevinmcmurtrie** [reply]
You're underestimating how evil Google and Apple are.  Google isn't fighting microSD cards, they're fighting all local processing that competes with their cloud services.  Despite all the improvements in flash chip stacking, phones have spent years stuck at 128/256/512 GB.  Features that allow cellphones to operate more like personal computers are being whittled away in the name of "security."

I terminated my subscription since there's no visible progress on a fix or workaround.  Despite having a recent $1200 phone, there's only crap 256 GB of built-in storage.

**Reply by tom-ch1** [reply]
> I think we should not try with this API cause it doesn't look designed for us. 

@vshcherb : I understand which option is not a valid solution from your perspective. What I don't understand is which option your proposed solution to this issue is. Can you share it with us please?

**Comment by kevinmcmurtrie** [comment]
I think a reasonable solution would be using MANAGE_EXTERNAL_STORAGE on the F-Droid version.  That's what other large-file apps are already doing.  That keeps the codebases in sync, keeps the app working, and helps punish Google for their subterfuge.

I'm going to cancel my subscription if I have to use 4.0.9 forever.  I've tried a lot of tricks from a lot of different web sites and none of them work on phones running stock Android. The results ranged from doing nothing to causing file corruption and crashes.

**Comment by rom4nb** [comment]
I just updated my Android 11 and now OsmAnd+ cannot access the external storage anymore. This is quite critical because I have hardly space on internal storage and need to use the SD Card storage. 

Is this connected to the EXTERNAL_STORAGE setting discussed here? Is there some way to fix it?

**Reply by kevinmcmurtrie** [reply]
Use OsmAnd~ 4.0.9 from F-Droid.  App permissions aren't usable in newer versions.

**Reply by tommycrock** [reply]
Or use 
SD_card/Android/media/net.osmand.plus/
See #12982

**Reply by kevinmcmurtrie** [reply]
That doesn't improve performance - it will be so slow that OsmAnd might appear to be crashed.  Stock android doesn't even have read-after-write consistency for media and Download folders so there will be a lot of spurious I/O errors.

**Reply by tommycrock** [reply]
It improves performance from not being able to access it.

**Reply by vanMacG** [reply]
> kevinmcmurtrie
>
> Use OsmAnd~ 4.0.9 from F-Droid. App permissions aren't usable in newer versions.

I downgraded to 4.0.9 but it makes no difference. Issue is still the same.

**Reply by kevinmcmurtrie** [reply]
In 4.0.9 you can grant storage access then move the storage to something on the microSD card other than the /Android/ directory.  This will make it fast and visible to other apps.  Example: ` /storage/1234-ABCD/OsmAnd`

**Reply by tom-ch1** [reply]
> Or use SD_card/Android/media/net.osmand.plus/ See #12982

I can confirm that this works, but I can also confirm that it's so slow that it is not a valid solution to the issue.
I also confirm that OsmAnd~ 4.0.9 from F-Droid and a storage location outside the Android directory (eg. /storage/1234-ABCD/OsmAnd/ not only works, but is also much faster. Thanks @kevinmcmurtrie for that hint!

**Comment by rom4nb** [comment]
Thanks guys, with the path SD_card/Android/media/net.osmand.plus it is working for me again and I could download my offline maps and use them. Thank you very much!

Am 2. Mai 2022 00:17:50 MESZ schrieb tommycrock ***@***.***>:

**Comment by sonora** [comment]
Sadly, @kevinmcmurtrie seems right observing "_Google isn't fighting microSD cards, they're fighting all local processing that competes with their cloud services._". 

This means: Rather than continuing to progress the world's online and/or google-controlled IT functionality, which would objectively be beneficial and fair, Google may now effectivley be fighting this planet's offline IT options and/or functionality not controlled by Google. And as such of course prominently the very purpose and functionality of OsmAnd itself!

I vote for us keeping up the fight here as long as we can, because
1. OsmAnd has a prime and obvious use case illustrating that the world also needs performant offline functionality and processing.
2. OsmAnd can prove like few others that this can, now could, be solved successfully and efficiently.
3. There should be a fight for a world with also offline IT options and/or with IT functionality not requiring always-on or semi-always-on services controlled by the big tech corps.
4. OsmAnd via it's expertise, use cases, and userbase must be part of that fight.
5. For OsmAnd this entire storyline is existential!

If we give up the battle over this new storage/performance issue, we already surrender some users and use cases. A big step towards folding our cards entirely, and one more nail in the casket of the world losing functioning and performant offline IT options.

Hence I vote for, as long as we possibly can, somehow and somewhere publishing an alternate and fully functional OsmAnd version that still works as intended, both to prove this to the world, and as a service to our subscribers.

Implementation-wise it could be done via MANAGE_EXTERNAL_STORAGE or via using targetSDK=29.  (Unfortunately, a switch like we used in our solution using Google's or the Android location API seems less feasible here...).

**Reply by vshcherb** [reply]
They fight against SDCard cause it allows user access to the files. So every program which allows to do that and every method will be likely banned in future. 
Unfortunately Google doesn't provide yet alternatively to Apple handy app Files and API to it, so every program with request could be accessed to the folder and this folder could be synced with any cloud or via Bluetooth / WiFi with any device in the network. 

I foresee that in nearest future even developers won't have access to the files and that would probably lead to a better security in the end. 

We plan to restructure app and provide every custom folder access, unfortunately it would require quite a change in UI / code structure etc and unlikely will work with Maps cause for Maps we need direct quick access to random parts of the file (API is very slow today to do that) but it could work for everything else where we need streaming GPX, Favorites, Audio / Video notes, ...

**Reply by kevinmcmurtrie** [reply]
The Google Play Store restrictions don't apply if you don't use Play Store.  Native access to microSD cards is still possible.  There's F-Droid, Amazon, an ordinary APK release, and other stores.

I don't think it's bad to send power users to use a different app store for a different assembly of the app.  I'm already using F-Droid for a few apps that have jumped off Play Store.

**Reply by tom-ch1** [reply]
> I foresee that in nearest future even developers won't have access to the files and that would probably lead to a better security in the end.

I think you might lose a substantial portion of your user base if this issue is going to last for much longer. The issue could be solved so easily by just declaring MANAGE_EXTERNAL_STORAGE for the F-Droid version. Could you do that for us until a more valid solution pops up?
Most gratefully, Tom

**Comment by sonora** [comment]
The Digital Markets Act (DMA), soon to be ratified by the European Union, has the potential of solving such issues, because one of the implications is that the use of alternative app stores must be supported by all manufacturers. While this has a majpr impact on Apple, one may speculate that it may also affect Android in the sense that Google Play Store mandates may lose some of their absoluteness.

This will open the opportunity to e.g. sell OsmAnd compilations targeting SDK 29, which as of today would solve all issues around file visibility, file access performance, and free folder selection, for all current versions of Android.

**Comment by ghost** [comment]
Well, I see I'm not the only one. I took photos with the OsmAnd plugin while hiking, impossible to recover the photos. Whether on internal or external storage, unable to recover files.

https://github.com/osmandapp/OsmAnd/discussions/14705

I have a Google Pixel 6 with Android 12 and downloaded the app from F-Droid. I tested several file manager, but none is able to recover the photos. I've been tearing my hair out with this for 2 weeks.

I don't understand why the photos are not showing up in the Google Photos gallery. Telegram, Signal, ... very confidential applications are able to create a folder in the gallery, why OsmAnd is so complicated. You will lose a lot of users if there is no solution. Currently the Photo plugin is unusable.

**Reply by sonora** [reply]
https://github.com/osmandapp/OsmAnd/issues/13254#issuecomment-984661388

**Comment by LucasBS1** [comment]
I used to put my maps (manually select) at /storage/emulated/0/Maps.
Now I can't

Is there a way to override this ?

I have a LSPosed module called NoStorageRestrict that does Miracles on every other app (EXSDCard Acess Enabler is obsolete on Android 11)

But OSMAND decided to be different. When I set the directory, it attempts to copy to it, and fails.

**Reply by anaxonda** [reply]
Not sure about the google-play version, at least the f-droid version has the possibility to grant "all files" permission that allows me to place maps in custom folder in my external sd card.

See https://github.com/osmandapp/OsmAnd/issues/13943#issuecomment-1200347641

**Reply by mariush444** [reply]
/storage/emulated/0/Maps can't works AFAIK.
You can put your map folder inside /storage/emulated/0/Android/media/ folder

It works for sure with osmand play and f-droid

**Reply by khimaros** [reply]
this does not work for me at all. actually, in attempting to use `/storage/emulated/0/media/osmand/` as the manually specified directory, i lost all of my stored tracks and favorites. take a backup before trying this.

**Reply by mariush444** [reply]
It works for Android 8-11. Many users confirmed that. I don't know about A12-13.
Please notice the path exactly.
Data loss during moving them from one directory to the another is known osmand's "feature". (ironic)
Backup is always welcome.  
I lost data twice :( before I learn my lazy nature painfully.
![Screenshot_1](https://user-images.githubusercontent.com/66887280/221869523-b8fe6250-a61c-4e52-a7c9-b8a052b786f8.jpg)
![Screenshot_2](https://user-images.githubusercontent.com/66887280/221869544-cdd6dc2c-7a34-495f-bdbe-5d7f00304b3b.jpg)

**Reply by khimaros** [reply]
i'm running CalyxOS 4.6.0 which is based on Android 13. i wonder if this was caused by the "Pause app activity if unused" and permission cleaners. maybe OSMAnd~ asked for filesystem access the first time i launched it, but since then the filesystem permission was removed.

**Reply by khimaros** [reply]
this is what fixed it for me:

- export all data from settings
- uninstall OSMAnd~ and do not keep data
- reinstall OSMAnd~ from F-Droid
- accept the permissions request for filesystem
- change storage setting
- set to "Manually Specified"
- select "Change" to change the path
- set path to `/storage/emulated/0/Android/media/net.osmand.plus/`
- save and follow remainder of wizard
- go to settings and reimport your data, replace all

in other words, the subfolder in `media` must be the package name `net.osmand.plus` and not simply `osmand`

**Reply by kevinmcmurtrie** [reply]
That's a completely broken directory in **stock** Android.  Other versions of Android may be OK, depending on how much of the Google codebase has been removed.  The Downloads directory is the same.  It's a FUSE mount to an Android app that tracks individual file ownership.  It's slow and buggy as hell.  Asking OsmAnd to move a large number of files there will cause errors, data loss, hanging, and maybe even reboot the phone.

You must be outside of the `Android` directory on stock Android.  I don't know how that's set up in Android 13 since Google is still trying hard to break competition to their Cloud Storage product.  I got mine working with the top-level directory of a microSD card when I had Android 12 by manually granting permissions for management of all files.  It's still working in Android 13 even though that permission is gone.

**Reply by contrapunctus-1** [reply]
@khimaros

> this is what fixed it for me:
>
>     * export all data from settings
>     * uninstall OSMAnd~ and do not keep data
>     * reinstall OSMAnd~ from F-Droid
>     * accept the permissions request for filesystem
>     * change storage setting
>     * set to "Manually Specified"
>     * select "Change" to change the path
>     * set path to `/storage/emulated/0/Android/media/net.osmand.plus/`
>     * save and follow remainder of wizard
>     * go to settings and reimport your data, replace all
>
> in other words, the subfolder in `media` must be the package name `net.osmand.plus` and not simply `osmand`

Thanks for writing this out, this is the only solution that worked for me on Android 13. The only difference from your steps is that OsmAnd+ (installed from F-Droid) did not ask me for filesystem permission, so I went into "app info" and granted it there before I changed the OsmAnd storage directory.

With that, after hours and hours of looking around and trying different workarounds, I'm finally able to synchronize OsmAnd's data using Syncthing as I was able to on my previous (Android 9) device. Downloaded maps and recorded tracks were synchronized, although settings were either not synchronized or are not exposed by OsmAnd (I'm not sure which) - I had to export/import them.

**Comment by sonora** [comment]
@kevinmcmurtrie  et al. It will continue to work "as usual" only if the installation file you use for OsmAnd in its manifest declares "All files access".

For apps published in the Google Play store this is limited to file manager type apps, and subject to a special approval. In that sense this is related to  #16059.

**Comment by mariush444** [comment]
I'm not java/google programmer but oruxmap's developer found a trick how to force A11 (maybe A12-13 too) limitation.

Before installation apk from gplay they install small apk from the website. After that new instalation of main apk have access to folders. Maybe a java dev. can create apk like this for osmand?

(GP is one of the version of oruxmap)
https://www.oruxmaps.com/cs/en/blog

```
Another option, it is a trick
1.-Unisntall the GP version.
2.-Install and open this small apk:[OruxMapsGP0.0.1.apk](https://oruxmaps.org/OruxMapsGP0.0.1.apk) It is an empty apk, it allows you, when updating to the real GP version, to keep permissions about access to the internal storage, like the old versions.
3.-Finally upgrade to the GP version from Google Play. IF THE APP ASKS YOU TO MIGRATE TO ANDROID 11 STORAGE SAY 'MAYBE LATER'.
After the upgrade, the app should be able to access all the folders, including the old from root SD or the root internal storage.
Remember do not migrate to Android 11!
And if you have to uninstall the app, you have to do the same procedure if you want the app to access the old folders.
```
